sockeye/beam_search.py

```diff
@@ -495,13 +529,23 @@ def forward(self, best_hyp_indices, *states):
 def _get_vocab_slice_ids(restrict_lexicon: Optional[lexicon.TopKLexicon],
                          source_words: pt.Tensor,
                          eos_id: int,
-                         beam_size: int) -> Tuple[pt.Tensor, int]:
+                         beam_size: int,
+                         target_prefix: Optional[pt.Tensor] = None) -> Tuple[pt.Tensor, int]:
     device = source_words.device
-    vocab_slice_ids = restrict_lexicon.get_trg_ids(source_words.cpu().int().numpy())
-    # Pad to a multiple of 8.
-    vocab_slice_ids = pt.nn.functional.pad(pt.tensor(vocab_slice_ids, device=source_words.device, dtype=pt.int64),  # type: ignore
-                                           pad=(0, 7 - ((vocab_slice_ids.size - 1) % 8)),
-                                           mode='constant', value=eos_id)
+    vocab_slice_ids_np = restrict_lexicon.get_trg_ids(source_words.cpu().int().numpy()) # type: ignore
+    vocab_slice_ids = None  # type: ignore
+    if target_prefix is not None:
+        # Ensuring that target prefix ids are part of vocab_slice_ids
+        vocab_slice_ids = pt.unique(pt.concat([pt.tensor(vocab_slice_ids_np, device=device, dtype=pt.int64), pt.flatten(target_prefix).type(pt.int64)], -1))
+        # Pad to a multiple of 8.
+        vocab_slice_ids = pt.nn.functional.pad(vocab_slice_ids.clone().detach(),
```

Comment:
just curious: why do you need the `clone()` and `detach()`?

Reply:
But why do you need to copy in the first place? You are not calling `pt.tensor` here. 
In any case, I think you could simplify this whole block to:
```
vocab_slice_ids = pt.tensor(vocab_slice_ids_np, device=device, dtype=pt.int64)
if target_prefix is not None:
    # Ensuring that target prefix ids are part of vocab_slice_ids
     vocab_slice_ids = pt.concat([vocab_slice_ids, target_prefix.flatten().type(pt.int64)], -1).unique()
# Pad to a multiple of 8.
vocab_slice_ids = pt.nn.functional.pad(vocab_slice_ids, pad=(0, 7 - ((vocab_slice_ids_np.size - 1) % 8)), mode='constant', value=eos_id)
```

Reply:
Haha silly me. Done. Very true and thanks Felix!

Reply:
thanks! I still think you can also avoid the duplicate call to `pad` as suggested above. After that I think this PR looks good to me! :)

Reply:
Yes, your code is very succinct. Done, thanks!